src/core/service.c

```diff
@@ -1401,6 +1401,31 @@ static bool service_exec_needs_notify_socket(Service *s, ExecFlags flags) {
         return s->notify_access != NOTIFY_NONE;
 }
 
+static Service *service_get_on_failure_of_dep(Service *s) {
+        assert(s);
+
+        /* If this service had a single reverse OnFailure= dependency of type
+         * UNIT_SERVICE which has failed, return that dependency. */
+
+        Unit *u = UNIT(s);
+        Unit *trigger;
+        Hashmap *deps;
+        void *v;
+
+        deps = unit_get_dependencies(u, UNIT_ON_FAILURE_OF);
+        if (deps && hashmap_size(deps) == 1) {
+                HASHMAP_FOREACH_KEY(v, trigger, deps) {
+                        if (trigger->type == UNIT_SERVICE) {
+                                Service *trigger_service = SERVICE(trigger);
+                                if (!IN_SET(trigger_service->result, SERVICE_SUCCESS, SERVICE_SKIP_CONDITION))
+                                        return trigger_service;
+                        }
```

Comment:
please also check if the unit is still down.

also, we usually prefer "early exit" over "deep nesting". i.e. when you notice that something isn't qualfying for what you do then do "return" or "continue" to leave the function early or the loop iteration early. That way you can keep indentation levels pretty short and the primary codepaths are the ones are somewhat linear in the function

Reply:
Will fix this in the next iteration. Though there is another problem here, In the case where we want to pass some status through to an OnSuccess= invoked service, the triggering service has already been collected and thus the service status data has all been freed. So, I figure I have the following options:

1) Restrict to supporting passing status through to OnFailure= dependencies which since the triggering service is inactive is not collected.
2) Somehow move collection of the successful service until after it's OnSuccess= dependency has been exec'd.
3) ?

Reply:
We have internal deps of type "References" for units to keep other units pinned. This shouldbe synthesized here too

Reply:
(or more specifically OnSucessOf= should contain the "REFERENCES" atom)

Reply:
Apologies for the delay on changes, latest changes work for both OnFailure= and OnSuccess= cases, though I had to manually add the REFERENCES/REFERENCED_BY dependency so may not have got this totally right.

OnSuccess= case:


```
root@localhost:~# systemctl status failer-handler --no-pager -l
○ failer-handler.service - Failer handler
     Loaded: loaded (/usr/lib/systemd/system/failer-handler.service; static)
     Active: inactive (dead)

Nov 17 17:11:05 localhost systemd[1]: Starting Failer handler...
Nov 17 17:11:05 localhost echo[471]: MONITOR_SERVICE_RESULT=success, MONITOR_EXIT_CODE=exited, MONITOR_EXIT_STATUS=0, MONITOR_INVOCATION_ID=25289a891c7b4c80a440b367bb1eed21, MONITOR_UNIT=failer.service
Nov 17 17:11:05 localhost systemd[1]: failer-handler.service: Deactivated successfully.
Nov 17 17:11:05 localhost systemd[1]: Finished Failer handler.
root@localhost:~# 

```

OnFailure= case:

```
root@localhost:~# systemctl status failer-handler --no-pager -l
○ failer-handler.service - Failer handler
     Loaded: loaded (/usr/lib/systemd/system/failer-handler.service; static)
     Active: inactive (dead) since Wed 2021-11-17 17:13:16 GMT; 2s ago
    Process: 491 ExecStart=/bin/echo MONITOR_SERVICE_RESULT=${MONITOR_SERVICE_RESULT}, MONITOR_EXIT_CODE=${MONITOR_EXIT_CODE}, MONITOR_EXIT_STATUS=${MONITOR_EXIT_STATUS}, MONITOR_INVOCATION_ID=${MONITOR_INVOCATION_ID}, MONITOR_UNIT=${MONITOR_UNIT} (code=exited, status=0/SUCCESS)
   Main PID: 491 (code=exited, status=0/SUCCESS)
        CPU: 1ms

Nov 17 17:13:16 localhost systemd[1]: Starting Failer handler...
Nov 17 17:13:16 localhost echo[491]: MONITOR_SERVICE_RESULT=exit-code, MONITOR_EXIT_CODE=exited, MONITOR_EXIT_STATUS=1, MONITOR_INVOCATION_ID=ec20674cf5cb4a9da7172edd2aabd185, MONITOR_UNIT=failer.service
Nov 17 17:13:16 localhost systemd[1]: failer-handler.service: Deactivated successfully.
Nov 17 17:13:16 localhost systemd[1]: Finished Failer handler.
root@localhost:~# 
```

Reply:
oops, need a rebase.

Reply:
Fixed that mess, though now there are some spurious tags on the PR :(